kanidmd/src/lib/plugins/oauth2.rs

```diff
@@ -0,0 +1,139 @@
+use crate::event::{CreateEvent, ModifyEvent};
+use crate::plugins::Plugin;
+use crate::prelude::*;
+use crate::utils::password_from_random;
+
+lazy_static! {
+    static ref CLASS_OAUTH2_BASIC: PartialValue =
+        PartialValue::new_class("oauth2_resource_server_basic");
+}
+
+pub struct Oauth2Secrets {}
+
+macro_rules! oauth2_transform {
+    (
+        $au:expr,
+        $e:expr
+    ) => {{
+        if $e.attribute_equality("class", &CLASS_OAUTH2_BASIC) {
+            if !$e.attribute_pres("oauth2_rs_basic_secret") {
+                lsecurity!($au, "regenerating oauth2 basic secret");
+                let v = Value::new_utf8(password_from_random());
+                $e.add_ava("oauth2_rs_basic_secret", v);
+            }
+            if !$e.attribute_pres("oauth2_rs_basic_token_key") {
+                lsecurity!($au, "regenerating oauth2 token key");
+                let k = fernet::Fernet::generate_key();
+                let v = Value::new_secret_str(&k);
+                $e.add_ava("oauth2_rs_basic_token_key", v);
+            }
+        }
+        Ok(())
```

Comment:
Does this just always return `Ok(())`? Or can `lsecurity!` return `Err(OperationError)`?

Reply:
In the future it *may* return errors if fields are not present or something goes wrong in this transformation.

Reply:
I'm starting to see a theme of future expandability here 😅

Reply:
Indeed :)